relayer/src/link/relay_path.rs

```diff
@@ -693,103 +734,281 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
         self.recv_packet_acknowledged_on_src(&rp.packet)
     }
 
-    /// Returns `true` if the delay for this relaying path is zero.
-    /// Conversely, returns `false` if the delay is non-zero.
-    pub fn zero_delay(&self) -> bool {
-        self.channel.connection_delay == ZERO_DURATION
+    /// Returns `true` if the connection delay for this relaying path is non-zero.
+    /// Conversely, returns `false` if the delay is zero.
+    #[inline]
+    fn has_delay(&self) -> bool {
+        self.channel.connection_delay != ZERO_DURATION
     }
 
-    /// Handles updating the client on the destination chain
-    fn update_client_dst(
+    /// Returns `true` iff the connection delay for this relaying path is non-zero and `op_data`
+    /// contains packet messages.
+    #[inline]
+    pub fn conn_delay_needed(&self, op_data: &OperationalData) -> bool {
+        self.has_delay() && op_data.has_packet_msgs()
+    }
+
+    #[inline]
+    fn conn_time_delay_elapsed(
         &self,
-        src_chain_height: Height,
-        tracking_id: &str,
-    ) -> Result<(), LinkError> {
-        // Handle the update on the destination chain
-        // Check if a consensus state at update_height exists on destination chain already
-        if self
-            .dst_chain()
-            .proven_client_consensus(self.dst_client_id(), src_chain_height, Height::zero())
-            .is_ok()
-        {
-            return Ok(());
+        op: &OperationalData,
+        chain_time: Instant,
+    ) -> Result<(), Duration> {
+        if self.conn_delay_needed(op) {
+            // since chain time instant is relative to relayer's current time, it is possible that
+            // `op.scheduled_time` is later (by nano secs) than `chain_time`, hence the call to
+            // `saturating_duration_since()`.
+            let elapsed = chain_time.saturating_duration_since(op.scheduled_time);
+            if elapsed > self.channel.connection_delay {
+                Ok(())
+            } else {
+                Err(elapsed)
+            }
+        } else {
+            Ok(())
         }
+    }
 
-        let mut dst_err_ev = None;
-        for i in 0..MAX_RETRIES {
-            let dst_update = self.build_update_client_on_dst(src_chain_height)?;
-            info!(
-                "sending updateClient to client hosted on destination chain for height {} [try {}/{}]",
-                src_chain_height,
-                i + 1, MAX_RETRIES,
-            );
-
-            let tm = TrackedMsgs::new(dst_update, tracking_id);
+    #[inline]
+    fn conn_block_delay_elapsed(
```

Comment:
This method of conn delay calculation is actually Tendermint specific according to the spec. So IIUC, chains can decide to ignore conn-delay completely or use an impl of their choice. This is another thing we might want to consider when we're adding support for non-SDK chains.